core/src/test/java/com/datastax/oss/driver/internal/core/util/concurrent/ScheduledTimeoutCapturingTimer.java

```diff
@@ -0,0 +1,126 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.datastax.oss.driver.internal.core.util.concurrent;
+
+import static org.assertj.core.api.Assertions.fail;
+
+import io.netty.util.Timeout;
+import io.netty.util.Timer;
+import io.netty.util.TimerTask;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Implementation of Netty's {@link io.netty.util.Timer Timer} interface to capture scheduled {@link
+ * io.netty.util.Timeout Timeout} instead of running them, events so they can be run manually in
+ * tests.
+ */
+public class ScheduledTimeoutCapturingTimer implements Timer {
+
+  private final ArrayBlockingQueue<Timeout> timeoutQueue = new ArrayBlockingQueue<>(16);
+
+  @Override
+  public Timeout newTimeout(TimerTask task, long delay, TimeUnit unit) {
+    // delay and unit are not needed as the Timeout's TimerTask will be run manually
+    Timeout timeout = new _Timeout(task, this);
+    // add the timeout to the queue
+    timeoutQueue.add(timeout);
+    return timeout;
+  }
+
+  /**
+   * Retrieves the next scheduled Timeout. In tests, this will usually be a request timeout or a
+   * speculative execution. Tests will need be able to predict the ordering as it is not easy to
+   * tell from the returned Timeout itself.
+   */
+  public Timeout getNextTimeout() {
+    try {
+      Timeout t = timeoutQueue.poll(100, TimeUnit.MILLISECONDS);
+      return t;
+    } catch (InterruptedException ie) {
+      fail("Unexpected interruption", ie);
+      throw new AssertionError();
+    }
+  }
+
+  @Override
+  public Set<Timeout> stop() {
+    if (timeoutQueue.isEmpty()) {
+      return Collections.EMPTY_SET;
+    }
+    Set<Timeout> timeoutsRemaining = new HashSet<>(timeoutQueue.size());
+    for (Timeout t : timeoutQueue) {
+      if (t != null) {
+        t.cancel();
+        timeoutsRemaining.add(t);
+      }
+    }
+    return timeoutsRemaining;
+  }
+
+  /**
+   * Implementation of Netty's {@link io.netty.util.Timeout Timeout} interface. It is just a simple
+   * class that keeps track of the {@link io.netty.util.TimerTask TimerTask} and the {@link
+   * io.netty.util.Timer Timer} implementation that should only be used in tests. The intended use
+   * is to call the {@link io.netty.util.TimerTask#run(io.netty.util.Timeout) run()} method on the
+   * TimerTask when you want to execute the task (so you don't have to depend on a real timer).
+   *
+   * <p>Example:
+   *
+   * <p>// get the timeout Timeout t = retrieveTimeout(); // run the timeout task t.task().run(t);
+   */
+  private static class _Timeout implements Timeout {
```

Comment:
I know it's an internal class, but not a big fan of the underscore in the name. Could we use something more descriptive, like maybe `CapturedTimeout` or `MockTimeout`?